storage/innobase/handler/ha_innodb.cc

```diff
@@ -15904,13 +15904,42 @@ static int innobase_alter_encrypt_tablespace(handlerton *hton, THD *thd,
   /* If new tablespace definition says it's encrypted */
   bool is_new_encrypted = !Encryption::is_none(newenc.data());
 
-  if (!is_old_encrypted && is_new_encrypted) {
+  // It is only possible to mark tablespace to be skipped by encryption threads
+  // if it is not already encrypted.
+  if (!is_old_encrypted && !is_new_encrypted &&
+      alter_info->explicit_encryption) {
+    if (!fil_crypt_exclude_tablespace_from_rotation(space)) {
+      my_error(ER_EXCLUDE_ENCRYPTION_THREADS_RUNNING, MYF(0), space->name);
+      DBUG_RETURN(convert_error_code_to_mysql(DB_ERROR, 0, NULL));
+    }
+    DBUG_RETURN(error);
+  } else if (!is_old_encrypted && is_new_encrypted) {
+    // TODO: PS-5323 should disallow reaching this code if encryption threads
+    // are enabled. Thus it should be safe to remove the crypt_data here. Here
+    // we encrypt with MK, the tablespace was previously excluded from rotation.
+    if (space->crypt_data) {
+      ut_ad(space->crypt_data->type == CRYPT_SCHEME_UNENCRYPTED);
+      fil_space_destroy_crypt_data(&space->crypt_data);
+      space->crypt_data = nullptr;
```

Comment:
This might cause troubles. many places space->crypt_data is used for takin decisions (wihtout any protection) and here too, without any protection, you are setting it to nullptr

Prone to race condition IMO

Reply:
as discussed on slack: crypt_data can be only removed if encryption threads are disabled. Since crypt_data can be only access concurrently by encryption threads - this is race free.